File: frontend/src/types/report.ts
/** Report and history contracts from the backend.

A "report" is the full persisted artifact for one completed interview: the
candidate summary, the structured feedback, and the transcript. History is the
per-candidate list of interviews with their overall scores.
*/

export interface ReportMessage {
  id: string;
  role: string;
  content: string;
  metadata?: Record<string, unknown>;
  created_at: string;
}

export interface ReportCandidate {
  id: string;
  name: string;
  role: string;
  email?: string | null;
  strengths?: string[];
}

export interface ReportTopic {
  topic_id: string;
  title: string;
  average_score: number;
  strengths?: string[];
  improvements?: string[];
}

export interface AnswerReview {
  question_id: string;
  topic_id: string;
  topic_title: string;
  question: string;
  answer: string;
  score: number;
  rationale: string;
  verdict: string;
}

export interface ReportFeedback {
  id: string;
  session_id: string;
  overall_score: number;
  summary: string;
  strengths: string[];
  improvements: string[];
  topics: ReportTopic[];
  created_at: string;
  source: string;
}

export interface InterviewReport {
  session_id: string;
  candidate: ReportCandidate;
  feedback: ReportFeedback;
  completed_at?: string | null;
  messages?: ReportMessage[];
  answer_reviews?: AnswerReview[];
}

export interface InterviewHistoryItem {
  session_id: string;
  state: string;
  created_at: string;
  updated_at: string;
  completed_at?: string | null;
  overall_score?: number | null;
  summary: string;
}

export interface InterviewHistory {
  candidate_id: string;
  items: InterviewHistoryItem[];
  total: number;
}

File: backend/app/models/report.py
"""Interview history and report response contracts.

A "report" is the full persisted artifact for one completed interview: the
session, the candidate, the structured feedback, and the transcript. History is
the per-candidate list of interviews.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import Field

from app.models.common import BaseModelConfig, Id
from app.models.feedback import FeedbackRead
from app.models.session import InterviewState


class ReportMessage(BaseModelConfig):
    """One transcript entry embedded in a report."""

    id: Id
    role: str = Field(description="system | interviewer | candidate")
    content: str
    metadata: dict = Field(default_factory=dict)
    created_at: datetime


class ReportCandidate(BaseModelConfig):
    """Candidate summary embedded in a report."""

    id: Id
    name: str
    role: str = Field(default="")
    email: str | None = None
    strengths: list[str] = Field(default_factory=list)


class AnswerReview(BaseModelConfig):
    """Per-question answer verification embedded in a report.

    Pairs a graded question with the candidate's answer and the score the
    evaluator recorded for it.
    """

    question_id: str
    topic_id: str = Field(default="")
    topic_title: str = Field(default="")
    question: str
    answer: str = Field(default="")
    score: float = Field(ge=0.0, le=10.0)
    rationale: str = Field(default="")
    verdict: str = Field(
        description="'Very good' when score > 8, otherwise 'Needs improvement'."
    )


class ReportRead(BaseModelConfig):
    """The complete report for a single interview session."""

    session_id: Id
    candidate: ReportCandidate
    feedback: FeedbackRead
    completed_at: datetime | None = None
    messages: list[ReportMessage] = Field(default_factory=list)
    answer_reviews: list[AnswerReview] = Field(default_factory=list)


class InterviewHistoryItem(BaseModelConfig):
    """One row in a candidate's interview history."""

    session_id: Id
    state: InterviewState
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None = None
    overall_score: float | None = Field(
        default=None,
        description="Present when the session has a persisted feedback report.",
    )
    summary: str = Field(default="")


class InterviewHistory(BaseModelConfig):
    """The interview history for a single candidate."""

    candidate_id: Id
    items: list[InterviewHistoryItem]
    total: int
